Add AVIF support

adamsilverstein opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/51228
<!-- insert a link to the WordPress Trac ticket here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein committed:
Add AVIF support

adamsilverstein committed:
whitespace

adamsilverstein committed:
Add AVIF constants

adamsilverstein committed:
update magic numbers for avif detection

adamsilverstein committed:
Include AVIF in the displayable image types

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
phpcbf

adamsilverstein committed:
Apply phpcbf fixes to avif helper

adamsilverstein committed:
more phpcbf

adamsilverstein committed:
yoda conditionals, other small phpcs fixes

adamsilverstein committed:
Restore avif helper original and exclude from phpcs

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

# Conflicts:
#	phpcs.xml.dist

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
remove excessive whitespace

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
Add AVIF test images

adamsilverstein committed:
match other test image names

adamsilverstein committed:
one more test AVIF image

adamsilverstein committed:
Add avif to get_image_mime tests

adamsilverstein committed:
Add AVIF to wp_getimagesize tests

adamsilverstein committed:
Add tests for wp_get_avif_info

note: the file type is not determined by this function as with webp, but the sizes are returned correctly.

adamsilverstein committed:
Add AVIF to file_is_valid_image_positive tests

adamsilverstein committed:
Add AVIF to file_is_displayable_image_positive test

adamsilverstein committed:
Add additional magic file detection numbers for AVIF in wp_get_image_mime

adamsilverstein committed:
Correctly handle 0x0 sizes returned from `getimagesize`, using shim fallback for AVIFs

adamsilverstein committed:
Add test_resize_avif

adamsilverstein committed:
phpcs

adamsilverstein committed:
remove unused module

adamsilverstein committed:
temporarily log debug info for test runs

adamsilverstein committed:
more debug logging for tests

adamsilverstein committed:
Imagick editor - fall back to `wp_getimagesize`

adamsilverstein committed:
Improve doc block

adamsilverstein committed:
more resize logging

adamsilverstein committed:
remove image size logging

adamsilverstein committed:
remove error logging

adamsilverstein committed:
log and skip failing test

adamsilverstein committed:
remove fallback

adamsilverstein committed:
log mime type check in imagick

adamsilverstein committed:
better logging

adamsilverstein committed:
add logging of failed avif test

adamsilverstein committed:
improve logging

adamsilverstein committed:
improve logging

adamsilverstein committed:
class name for log

adamsilverstein committed:
log avifinfo results after resize

adamsilverstein committed:
skip if either h or w empty

adamsilverstein committed:
more logging

adamsilverstein committed:
more logging

adamsilverstein committed:
more logging

adamsilverstein committed:
log mime magics

adamsilverstein committed:
adjust empty logic, log after check

adamsilverstein committed:
log a mime check in test

adamsilverstein committed:
fix typo

adamsilverstein committed:
add additional magic for AVIF detection

adamsilverstein committed:
remove logging

adamsilverstein committed:
Update doc blocks, cleanup

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
add fallback comment

adamsilverstein committed:
improve libavifinfo doc block, link to source

adamsilverstein committed:
remove lossless AVIF image handling: not supported in PHP

adamsilverstein committed:
Simplify wp_get_avif_info and improve doc blocks

adamsilverstein committed:
Add todo for detection change

y-guyon commented on src/wp-includes/class-avif-info.php:
```diff
@@ -8,6 +8,8 @@
  * Media Patent License 1.0 was not distributed with this source code in the
  * PATENTS file, you can obtain it at www.aomedia.org/license/patent.
+ *
+ * Note: this class is from libavifinfo - https://aomedia.googlesource.com/libavifinfo/+/refs/heads/main/avifinfo.php
```
I suggest using a permalink such as https://aomedia.googlesource.com/libavifinfo/+/b496868f7c3fd17dfeeecc0364fe37e19edd548a/avifinfo.php or adding "at b496868" maybe

adamsilverstein replied:
added in b6b78d019283ad77dd57665aee2f8d00acf7f0af

y-guyon commented on src/js/_enqueues/vendor/thickbox/thickbox.js:
```diff
@@ -76,15 +76,16 @@ function tb_show(caption, url, imageGroup) {//function called when the user clic
 	   }
 
-	   var urlString = /\.jpg$|\.jpeg$|\.png$|\.gif$|\.bmp$|\.webp$/;
+	   var urlString = /\.jpg$|\.jpeg$|\.png$|\.gif$|\.bmp$|\.webp$|\.avif$/;
```
Note that MIME type and file extension [heif](https://mimetype.io/image/heif) could be technically valid, but as for supporting other major brands, it is not worth it.

Also see https://aomediacodec.github.io/av1-avif/v1.1.0.html#brands-overview on the topic.

adamsilverstein replied:
I think we would need to consider `.heif` files separately (are they technically AVIF files?)... also browser support is a prerequisite to supporting the format and so far only Safari support heif - https://caniuse.com/heif

y-guyon replied:
> I think we would need to consider `.heif` files separately

Is there a user need for that?

> are they technically AVIF files?

According to the specification I would say yes, without certitude. But I doubt any library generates them this way.  
It would be `heif` as a major_brand and `avif` as a compatible_brand.  
Anyway this is not even worth a discussion to be honest.

> so far only Safari support heif

I think you meant HEIC. I agree it is a bit confusing and finding HEIF associated with HEIC everywhere does not help.  
Also there is little interest for browsers to support the generic container HEIF without the specific underlying formats (AV1, HEVC). That is why there is AVIF or HEIC support but it won't mention MIAF, HEIF or ISOBMFF support because these pure containers cannot be rendered on browsers.  
CMS like WordPress on the other end could take advantage of the codec-agnostic image properties such as width/height defined directly in HEIF property `ispe` (as long as it is known that there is enough browser support for it). I assume WordPress does not care about pixel values but as you mentioned, it might handle resizing and stuff so my point is only for page layout purposes.

adamsilverstein replied:
> I think you meant HEIC. I agree it is a bit confusing and finding HEIF associated with HEIC everywhere does not help.

Ah right, I was confusing the two.

y-guyon commented on src/wp-includes/class-wp-image-editor-gd.php:
```diff
@@ -562,7 +578,13 @@ public function stream( $mime_type = null ) {
 					return imagewebp( $this->image, null, $this->get_quality() );
 				}
-				// Fall back to the default if webp isn't supported.
+				// Fall back to JPEG as the default.
```
Update comment

adamsilverstein replied:
Right, this would actually fall thru to the AVIF block which isn't right, fixing.

adamsilverstein replied:
Fixed in c0267710f74717ed29c278ba1ae1c72fabab3a27

y-guyon commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -256,6 +257,14 @@ protected function update_size( $width = null, $height = null ) {
 
+		// If we still don't have the image size, fall back to `wp_getimagesize`. This ensures AVIF images
+		// are properly sized without affecting previous `getImageGeometry` behavior.
+		if ( ! $width && ! $height && 'image/avif' === $this->mime_type ) {
```
Should it be `! $width || ! $height`?

adamsilverstein replied:
Probably, I had tried that in a previous commit then reverted it. In my testing (our unit tests run across a huge matrix of configuration) they were either both valid or both 0. Still, I agree that if either is 0 we should use the helper instead, so I'll change this back.

adamsilverstein replied:
updated in ddbd3440f50bc3dd06bd98c8a4cd4601a60558df

adamsilverstein replied:
Improved in ddbd3440f50bc3dd06bd98c8a4cd4601a60558df

y-guyon commented on src/wp-includes/functions.php:
```diff
@@ -3349,6 +3351,24 @@ function wp_get_image_mime( $file ) {
+		/**
+		 * Add AVIF fallback detection when image library doesn't support AVIF.
+		 *
+		 * @todo check the third 4-byte token "[major_brand]", for "avif" or "avis" .
```
AVIF spec inherits MIAF and HEIF so it relies on ISOBMFF.  
ISOBMFF files start with a `FileTypeBox` whose syntax begins with:
- 4-byte big-endian unsigned integer being the number of bytes used by the `FileTypeBox`, including this field
- 4-byte hard-coded literal string `"ftyp"`
- 4-byte string `"[major_brand]"`
- other fields

So I would check for:
- second 4-byte token being exactly `"ftyp"`
- third 4-byte token being exactly `"avif"` or `"avis"`
- unnecessary but the first three bytes are very likely to be 0 (no more than 60 compatible brands in `FileTypeBox`)

Section `4.3.1 File-type box Definition` of specification ISO/IEC 14496-12 can be cited and https://aomediacodec.github.io/av1-avif/v1.1.0.html#brands can be cited for the major brands that can be reasonably expected in an AVIF file (`avio` should not).

adamsilverstein replied:
Adopted this approach in 3ecb6825eeb5a8b224a115a07a19906fd457270a

adamsilverstein replied:
This worked well in my testing and simplified the logic here; thanks for the tip!

y-guyon commented on tests/phpunit/data/images/avif-transparent.avif:
This file looks broken in Chrome but GIMP opens it successfully. There are compliance issues according to https://gpac.github.io/ComplianceWarden-wasm/.

Here is a 128*128px gradient translucent still 12-bit AVIF generated with GIMP that opens in Chrome, feel free to use it or to generate another:
![avif-transparent2 avif](https://github.com/WordPress/wordpress-develop/assets/51164219/7eb3fd65-a645-402c-b513-36717996c6ce)

adamsilverstein replied:
Thanks, I'll swap the test image out - I struggled to find the best image for these tests.

adamsilverstein replied:
:( github turned that upload into a png in its attempt to be helpful.

y-guyon replied:
Right click > `Save image as...` > "img.avif" works on my end.

Note that to upload an AVIF file on GitHub, I renamed it to end with `.png`, but GitHub keeps the file as is without transcoding. It is just the displayed file extension that is wrong.

adamsilverstein replied:
> Note that to upload an AVIF file on GitHub, I renamed it to end with .png, but GitHub keeps the file as is without transcoding. It is just the displayed file extension that is wrong.

Oh, tricky! So the ".png" file I downloaded is actually an avif... the file extension fooled me.

y-guyon commented on tests/phpunit/tests/image/editor.php:
```diff
@@ -363,4 +363,88 @@ public function data_wp_get_webp_info() {
+					'width'        => 128,
+					'height'       => 128,
+					'bit_depth'    => 12,
+					'num_channels' => 3,
```
This should be 4 if the AVIF is translucent.

adamsilverstein replied:
Apparently the image I used to the tests wasn't actually transparent. Once I correct the image, this test should fail - then I will fix the test.

adamsilverstein replied:
I swapped in the gradient image you shared and the test still passes with 3 channels at least locally, running the full matrix on the PR now. I added a few more test images from libavif and also made sure the tests for `wp_get_avif_info` always run regardless of server support for AVIF (also for the WebP tests).

> This should be 4 if the AVIF is translucent.

I think the image is right, so maybe there is something wrong with the PHP code? This isn't critical since we don't actually use the `bit_depth` anywhere, however it would be good to dig in further to see why this isn't working.

y-guyon replied:
> I think the image is right, so maybe there is something wrong with the PHP code?

Indeed, I was pretty confident with the logic because I copied it from the C implementation which is tested and fuzzed, but my alpha string comparison was off in PHP. See the [fix](https://aomedia-review.googlesource.com/c/libavifinfo/+/186521).

I suggest these two changes in your PR:
- Add `\0` in `src/wp-includes/class-avif-info.php:453`
- Update from `b496868` to `f509487` in `src/wp-includes/class-avif-info.php:12`

adamsilverstein replied:
> alpha string comparison was off in PHP

Well, the good news is running the tests here caught the bug! I will apply the changes you suggested and update the tests.

adamsilverstein replied:
Fixed in [`b7c193c` (#4612)](https://github.com/WordPress/wordpress-develop/pull/4612/commits/b7c193ca55470bc50eac13ee3f05c7d17ead0d87)

adamsilverstein committed:
WP_Image_Editor_GD:stream - improve mime detection and fallback

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
WP_Image_Editor_Imagick::update_size - improve logic for avif fallback

adamsilverstein committed:
Update header based AVIF detection

adamsilverstein committed:
Add commit hash to libavifinfo reference

adamsilverstein committed:
revert creation of additional header variable

adamsilverstein committed:
Doc block cleanup.

y-guyon commented on src/wp-includes/functions.php:
```diff
@@ -3358,14 +3358,15 @@ function wp_get_image_mime( $file ) {
 		 * @todo check the third 4-byte token "[major_brand]", for "avif" or "avis" .
```
TODO can be removed

I also recommend mentioning which specifications to look into for reference here

adamsilverstein replied:
replaced the todo with a link to the specification from your previous comment.

adamsilverstein committed:
replace avif-transparent test image

adamsilverstein committed:
Replace @todo comment with link to specification

adamsilverstein committed:
Always test `wp_get_avif_info` regardless of server AVIF support

adamsilverstein committed:
Add additional AVIF test images, remove one unused image

adamsilverstein committed:
Always test wp_get_webp_info regardless of server support for webp

this function is designed to work regardless of server support for WebP so let’s test it that way

adamsilverstein committed:
Update avif-info doc block to indicate it will no longer be required after PHP 8.2 is the minimum version for WordPress

adamsilverstein committed:
Correct bit_depth detection bug in avif-info class, adjust tests to match

y-guyon commented on src/wp-includes/functions.php:
```diff
@@ -3355,17 +3355,18 @@ function wp_get_image_mime( $file ) {
-		 * @todo check the third 4-byte token "[major_brand]", for "avif" or "avis" .
-		 *
+		 * Detection based on section 4.3.1 File-type box Definition of the ISO/IEC 14496-12
+		 * specification, https://see aomediacodec.github.io/av1-avif/v1.1.0.html#brands.
```
Nit:

```diff
Detection based on section 4.3.1 File-type box Definition of the ISO/IEC 14496-12
-specification, https://see aomediacodec.github.io/av1-avif/v1.1.0.html#brands.
+specification, and AV1-AVIF spec https://see aomediacodec.github.io/av1-avif/v1.1.0.html#brands.
```

adamsilverstein replied:
looks good, also fixing the link which has "see" inside the https reference

mukeshpanchal27 commented on tests/phpunit/tests/image/editor.php:
```diff
@@ -363,4 +357,100 @@ public function data_wp_get_webp_info() {
+	 * Test wp_get_avif_info.
+	 *
+	 * @ticket 51228
+	 * @dataProvider data_wp_get_avif_info
+	 *
```
```suggestion
	 *
	 * @dataProvider data_wp_get_avif_info
	 *
	 * @param string $file     The path to the AVIF file for testing.
	 * @param array  $expected The expected AVIF file information.
```

adamsilverstein replied:
great, thanks!

mukeshpanchal27 commented on src/wp-includes/media.php:
```diff
@@ -5541,10 +5552,78 @@ function wp_getimagesize( $filename, array &$image_info = null ) {
+ * }
+ */
+function wp_get_avif_info( $filename ) {
+	$width  = false;
+	$height = false;
+	$type   = false;
+
```
```suggestion
```

Remove unused local variables.

adamsilverstein committed:
Update src/wp-includes/media.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

adamsilverstein committed:
Update tests/phpunit/tests/image/editor.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

adamsilverstein committed:
Update doc block for AVIF mime detection

adamsilverstein committed:
Clean up return from `wp_get_avif_info`

adamsilverstein committed:
Correct tests for `wp_get_avif_info`

adamsilverstein commented:
@mukeshpanchal27 - note that it looks like tests broke in [`49f4278` (#4612)](https://github.com/WordPress/wordpress-develop/pull/4612/commits/49f4278b61cb49b678d5a6f69a29ede2db52e62b). those variables were used on the next line where the line `return compact( 'width', 'height', 'type' );` uses a weird syntax to reference them.

Anyway, that function was poorly structured and had an inconsistent return structure; it already had a default return value set later in the function so I moved that up and return it when the mime type isn't avif in  [`b492dbf` (#4612)](https://github.com/WordPress/wordpress-develop/pull/4612/commits/b492dbff3a256103b6c7b56387f4587109eb6d4e)

adamsilverstein committed:
wp_get_avif_info - test fix for default return

mukeshpanchal27 commented on phpcs.xml.dist:
```diff
@@ -59,6 +59,8 @@
 	<exclude-pattern>/src/wp-includes/class-requests\.php</exclude-pattern>
 	<exclude-pattern>/src/wp-includes/class-simplepie\.php</exclude-pattern>
 	<exclude-pattern>/src/wp-includes/class-snoopy\.php</exclude-pattern>
+	<exclude-pattern>/src/wp-includes/class-wp-block-parser\.php</exclude-pattern>
```
```suggestion
```

Why is this file being excluded?

adamsilverstein replied:
hmm, good question! I'm not sure and it looks extraneous, I will try removing that line.

mukeshpanchal27 commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -256,6 +257,14 @@ protected function update_size( $width = null, $height = null ) {
 			$height = $size['height'];
 		}
 
+		// If we still don't have the image size, fall back to `wp_getimagesize`. This ensures AVIF images
+		// are properly sized without affecting previous `getImageGeometry` behavior.
```
```suggestion
		/*
		 * If we still don't have the image size, fall back to `wp_getimagesize`. This ensures AVIF images
		 * are properly sized without affecting previous `getImageGeometry` behavior.
		 */
```
Nit.

mukeshpanchal27 commented on src/wp-includes/media.php:
```diff
@@ -5512,7 +5519,11 @@ function wp_getimagesize( $filename, array &$image_info = null ) {
+		! empty( $info ) &&
+		// Some PHP versions return 0x0 sizes from `getimagesize` for unrecognized image formats, including AVIFs.
+		! ( empty( $info[0] ) && empty( $info[1] ) )
+		) {
```
```suggestion
	) {
```

mukeshpanchal27 commented on src/wp-includes/media.php:
```diff
@@ -5541,10 +5552,75 @@ function wp_getimagesize( $filename, array &$image_info = null ) {
+ * @return array {
+ *    An array of AVIF image information.
+ *
+ *    @type int|false    $width        Image width on success, false on failure.
+ *    @type int|false    $height       Image height on success, false on failure.
+ *    @type int|false    $bit_depth    Image bit depth on success, false on failure.
+ *    @type int|false    $num_channels Image number of channels on success, false on failure.
```
```suggestion
 *    @type int|false $width        Image width on success, false on failure.
 *    @type int|false $height       Image height on success, false on failure.
 *    @type int|false $bit_depth    Image bit depth on success, false on failure.
 *    @type int|false $num_channels Image number of channels on success, false on failure.
```

mukeshpanchal27 reviewed (approved):
Thanks, @adamsilverstein, for the update. I've left some nit-pick feedback and suggestions. Other than that, looks good to me.

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein committed:
Update src/wp-includes/class-wp-image-editor-imagick.php

docblock improvement

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

adamsilverstein committed:
Update src/wp-includes/media.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

adamsilverstein committed:
Update src/wp-includes/media.php

docblock cleanup

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

adamsilverstein committed:
Update phpcs.xml.dist

remove inadvertent exclude path for phpcs

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

adamsilverstein commented:
> Thanks, @adamsilverstein, for the update. I've left some nit-pick feedback and suggestions. Other than that, looks good to me.

Thanks for the careful review and suggestions, I have accepted your suggestions.

adamsilverstein committed:
Merge branch 'trunk' into add/avif-support

adamsilverstein commented:
merged in https://core.trac.wordpress.org/changeset/57524
